libraries/AP_Stats/AP_Stats.cpp

```diff
@@ -150,12 +163,62 @@ void AP_Stats::set_flying(const bool is_flying)
         if (_flying_ms) {
             update_flighttime();
             update_runtime();
+            update_distance();
             flush();
         }
         _flying_ms = 0;
     }
 }
 
+void AP_Stats::update_distance()
+{
+    if (!_flying_ms) {
+        // Dont count unless flying
+        _last_position_valid = false;
+        return;
+    }
+
+    // Run at 1Hz to reduce how much GPS variance is counted as distance
+    const uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - _last_distance_update_ms < 1000) {
+        return;
+    }
+    _last_distance_update_ms = now_ms;
+
+    AP_AHRS &ahrs = AP::ahrs();
+    if (!ahrs.healthy()) {
+        _last_position_valid = false;
+        return;
+    }
+
+    Location current_loc;
+    if (!ahrs.get_location(current_loc)) {
+        _last_position_valid = false;
+        return;
+    }
+
+    if (_last_position_valid) {
+        float horiz_dist_m = _last_position.get_distance(current_loc);
+        float alt_diff_m = (current_loc.alt - _last_position.alt) * 0.01f;
+
+        // get_distance only calculates 2D distance, so account for altitude differences
+        float delta_m = sqrtf(sq(horiz_dist_m) + sq(alt_diff_m));
+
+        // Avoid counting small distances that are probably error
+        if (delta_m >= 0.5) {
+            // Dont blindly add large distances that are likely just gross error / spoofing
+            // Assumes vehicle traveling below mach ~0.9
+            if (delta_m > 300.0f) {
+                delta_m = 300.0f;
```

Comment:
I wonder if ardupilot has flown at supersonic speeds. Maybe make this a parameter? Constant?